sycl/doc/extensions/experimental/sycl_ext_oneapi_usm_raii.asciidoc

```diff
@@ -0,0 +1,135 @@
+= sycl_ext_oneapi_usm_raii
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:dpcpp: pass:[DPC++]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2025 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 9 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback.  Interfaces defined in this
+specification are implemented in {dpcpp}, but they are not finalized and may
+change incompatibly in future versions of {dpcpp} without prior notice.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+SYCL 2020 provides very C-like interfaces for USM allocation/deallocations
+routines. With in-order queues/USM and
+link:./sycl_ext_oneapi_enqueue_functions.asciidoc[sycl_ext_oneapi_enqueue_functions]
+as a future direction to write SYCL programs with as little overhead as possible
+we need usability improvements for the USM APIs that C++ programmers would
+expect.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_USM_RAII` to one of the values defined in
+the table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== USM Deleter
+
+This extension adds `usm_deleter` helper that can be used with `std::unique_ptr`
+or `std::shared_ptr` to manage deallocation of USM memory.
```

Comment:
Are there any restrictions on the type of USM allocation that can be used?

IIRC, `usm_allocator` cannot be used with `device` memory, because something inside the standard library assumes that the pointer can be dereferenced on the host. Is `usm_deleter` different because no code except the `operator()` is called?

Reply:
No limitations here, it just moves manual `sycl::free` to the automatic call from inside destructor on scope exit. I'll push an update clarifying this shortly.